receiver/promtailreceiver/config.go

```diff
@@ -0,0 +1,117 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package promtailreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/promtailreceiver"
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/clients/pkg/promtail/positions"
+	"github.com/grafana/loki/clients/pkg/promtail/scrapeconfig"
+	"github.com/grafana/loki/clients/pkg/promtail/targets/file"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.opentelemetry.io/collector/confmap"
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v2"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/adapter"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/promtailreceiver/internal"
+)
+
+// PromtailConfig defines configuration for the promtail receiver
+type PromtailConfig struct {
+	InputConfig        Config `mapstructure:",squash" yaml:",inline"`
+	adapter.BaseConfig `mapstructure:",squash" yaml:",inline"`
+}
+
+// NewConfig creates a new input config with default values
+func NewConfig() *Config {
+	return NewConfigWithID(operatorType)
+}
+
+// NewConfigWithID creates a new input config with default values
+func NewConfigWithID(operatorID string) *Config {
+	return &Config{
+		InputConfig: helper.NewInputConfig(operatorID, operatorType),
+	}
+}
+
+type PromtailInputConfig struct {
+	PositionsConfig positions.Config      `mapstructure:"positions,omitempty" yaml:"positions,omitempty"`
+	ScrapeConfig    []scrapeconfig.Config `mapstructure:"scrape_configs,omitempty" yaml:"scrape_configs,omitempty"`
+	TargetConfig    file.Config           `mapstructure:"target_config,omitempty" yaml:"target_config,omitempty"`
+}
+
+type Config struct {
+	helper.InputConfig `mapstructure:",squash" yaml:",inline"`
+	// Promtail input config is declared as nested key to allow custom unmarshalling
+	// confmap doesn't call custom unmarshal for the top structure
+	// see the implementation of unmarshalerHookFunc in https://github.com/open-telemetry/opentelemetry-collector:
+	// https://github.com/open-telemetry/opentelemetry-collector/blob/main/confmap/confmap.go#L274
```

Comment:
I'm not sure this is accurate. Many other components implement custom unmarshal functions for their top-level configs. I believe the top-level issue only applies when calling unmarshal directly on the struct, instead of allowing the hook func to do so.

In any case, if this ends up being necessary, please use a permalink to the code. It appears to already be slightly out of date as some changes were made to the file.

Reply:
Thank you for brining that. Now it makes sense to me. Actually custom unmarshall works fine when collector is running, but it isn't called in operatortest. 
It seems that operatortest calls unmarshall directly on the struct [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/bd43e7355cea53f08a35256dab812d9c0076a7cf/pkg/stanza/operator/operatortest/confmap.go#L80)

So yeah, the problem with top-level config was that operatortest failed, because it didn't call custom unmarshall method.

To move this PR forward I can bring back top-level config and remove tests that are using operatortest for now. Not sure how to fix the issue with operatortest right now, probably it is better to do in the separate PR

Reply:
Ah, I see. That's unfortunate. We can revisit after we have a solution along the lines of https://github.com/open-telemetry/opentelemetry-collector/pull/6236